ionic-team/stencil
# fix(compiler): make sure typesDir exist before writing to it

fixes #5091 / STENCIL-1024




## What is the current behavior?

Given the following Stencil config:

```ts
export const config: Config = {
  namespace: 'stencil-type-issue',
  outputTargets: [
    {
      type: 'dist',
      esmLoaderPath: '../loader',
    },
    {
      type: 'dist-custom-elements',
    },
  ],
  plugins: [
    sass()
  ],
  testing: {
    browserHeadless: "new",
  },
};
```

running a Stencil build causes the types directory `/dist/types` not being generated.

My investigations were as follows:
- having `{ type: 'dist-custom-elements' }` entry in `outputTargets` seems to cause the type generation to run twice with the same configurations
- this causes [`generateTypes`](https://github.com/ionic-team/stencil/blob/2b0727efdf6062445655725ad6923fb2cad4eaab/src/compiler/types/generate-types.ts#L15) to be called twice as well as the command to write [`/dist/types/components.d.ts`](https://github.com/ionic-team/stencil/blob/5d8387f0a1c9f860151d133598dabb22e2b1f911/src/compiler/types/generate-app-types.ts#L46-L48) with the option [`immediateWrite: true`](https://github.com/ionic-team/stencil/blob/5d8387f0a1c9f860151d133598dabb22e2b1f911/src/compiler/types/generate-app-types.ts#L47)
- adding a "sleep" right after has shown that actually nothing is being written to file even though `fs.writeFile` is called in [`/src/sys/node/node-sys.ts#L612-L614`](https://github.com/ionic-team/stencil/blob/4959295c24ec3effcc8d63a8305dffd6e07a617d/src/sys/node/node-sys.ts#L612-L614)

GitHub Issue Number: #5091

## What is the new behavior?

As seen in the change-set it turns out that we didn't handle the error provided by the `sys.writeFile`. Essentially the call to create the file failed without handling its error. Further investigations have shown that the directory for the file didn't exist even though we were calling [`ensureDir`](https://github.com/ionic-team/stencil/blob/db01eb840602f4c023e488aaf266795cee2dec57/src/compiler/sys/in-memory-fs.ts#L735) before. The race condition originates from the question how fast other output items were able to generate their types (I believe).

My suggestion to resolve this issue is to use the `recursive` flag for creating the directory when we are not working within memory. I am afraid calling `mkdir` multiple times, especially on `graceful-fs` rather than the Node.js package must have caused weird side effects in the event loop. During the investigations I literally was able to resolve it by adding a simple `console.log`.

An alternative solution would be to remove the `immediateWrite: true` option and have it write to the filesystem when everything file operation is being flushed which is likely a point in time the directory has been created already. However since  we are calling `generateAppTypes` also in [`runTsProgram`](https://github.com/ionic-team/stencil/blob/2a75b6501f4f76dad0d8fa8304af57be1c04eef1/src/compiler/transpile/run-program.ts#L111) I assume it is important to keep this option so rewrites are happening on watch mode!?

## Does this introduce a breaking change?




## Testing

See Jira ticket, apply the patch provided and reproduce the issue.

To test this change:

- checkout the branch
- build the project `npm run build`
- set your local Stencil dev environment as dependency in the reproduction project, e.g. `"@stencil/core": "file:/Users/christian.bromann/Sites/Ionic/projects/stencil"`
- run `./loop.sh` and see that the `types` dir is created every time

## Other information
